### PR TITLE
Support for formatting partial blocks of codes.

### PR DESCRIPTION
**Pull request by purpleP**
This as far as I've tested closes [https://github.com/google/yapf/issues/224](#224)
I've ran your tests (changing two of them which were checking formatting of bad input) they all green.
Also I've tested if manually on some inputs and it works fine.
One thing though. If the indentation was incorrect (not a multiple of four etc) yapf wouldn't fix it for partial inputs. Which is actually easy to fix. But me or someone else could add this later.

### CODE REVIEW

**Comment by purpleP** [comment]
I signed it!

**Comment by gwelymernans** [comment]
Thank you. Could you squash the patches together?

File: yapf/yapflib/py3compat.py [hunk 1]
```diff
@@ -83,3 +83,7 @@ class ConfigParser(configparser.ConfigParser):
 
     def read_file(self, fp, source=None):
       self.readfp(fp, filename=source)
+
+
+def indent(text, indent):
+    return '\n'.join(indent + l for l in text.split('\n')[:-1])
```
Comment (gwelymernans) [inline yapf/yapflib/py3compat.py]: How about this instead?

```
('\n'+indent).join(text.split('\n')[:-1])
```

File: yapf/__init__.py [hunk 2]
```diff
@@ -143,16 +150,26 @@ def main(argv):
...
-        py3compat.unicode('\n'.join(original_source) + '\n'),
+        py3compat.unicode(text_to_format),
         filename='<stdin>',
         style_config=style_config,
         lines=lines,
         verify=args.verify)
-    sys.stdout.write(reformatted_source)
+    out = reformatted_source
+    if indent_match:
+        out = indent(out, indent_match.group(0))
```
Comment (gwelymernans) [inline yapf/__init__.py]: Do you need a terminating `\n` here?

File: yapf/__init__.py [hunk 3]
```diff
@@ -143,16 +150,26 @@ def main(argv):
         original_source.append(py3compat.raw_input())
       except EOFError:
         break
+
+    original_source.append('')
+    text_to_format = '\n'.join(original_source)
```
Comment (gwelymernans) [inline yapf/__init__.py]: You need a terminating `\n` I think.

File: yapf/yapflib/py3compat.py (hunk 1 above)
Comment (gwelymernans) [inline yapf/yapflib/py3compat.py]: Also, why is this function in `py3compat`? That module is for Python 3 compatibility functions only.

File: yapf/__init__.py [hunk 4]
```diff
@@ -30,16 +30,23 @@
 import argparse
 import logging
 import os
-import sys
+from re import match
+
+try:
+    from textwrap import dedent, indent
+except ImportError:
+    from py3compat import indent
```
Comment (gwelymernans) [inline yapf/__init__.py]: Please move this to `py3compat.py`.

File: yapf/__init__.py [hunk 5]
```diff
@@ -30,16 +30,23 @@
 import argparse
 import logging
 import os
-import sys
+from re import match
+
+try:
+    from textwrap import dedent, indent
```
Comment (gwelymernans) [inline yapf/__init__.py]: I prefer not importing `dedent` from `textwrap`, but using `textwrap.dedent` when necessary.

**Comment by gwelymernans** [comment]
@purpleP Please squash all of your changes into one commit so that it's easier to review. Thanks! :-)

**Comment by purpleP** [comment]
My eyes hurt after this, so I'll do that later. However I see all changes like one commit and don't know how to squash it.
23 апр. 2016 г. 2:26 пользователь Bill Wendling <notifications@github.com> написал:@purpleP Please squash all of your changes into one commit so that it's easier to review. Thanks! :-)

**Comment by gwelymernans** [comment]
Basically, you just have to do `git rebase -i` and then mark all but the first commit as "squash". It's really best when there are several commits for a pull.

See more here:

  https://help.github.com/articles/about-git-rebase/

File: yapf/__init__.py [hunk 6]
```diff
@@ -143,11 +143,13 @@ def main(argv):
...
       except EOFError:
         break
+
+    original_source.append('')
     style_config = args.style
     if style_config is None and not args.no_local_style:
       style_config = file_resources.GetDefaultStyleForDir(os.getcwd())
     reformatted_source, changed = yapf_api.FormatCode(
-        py3compat.unicode('\n'.join(original_source) + '\n'),
+        py3compat.unicode('\n'.join(original_source)),
```
Comment (gwelymernans) [inline yapf/__init__.py]: I'm not sure why you added an empty string to `original_source` just to get the terminating newline instead of doing it here. Is it more efficient?

File: yapf/__init__.py [hunk 7]
```diff
@@ -210,13 +212,13 @@ def FormatFiles(filenames,
...
-      if has_change and reformatted_code is not None:
-        file_resources.WriteReformattedCode(filename, reformatted_code,
-                                            in_place, encoding)
       changed |= has_change
     except SyntaxError as e:
       e.filename = filename
       raise
+    if reformatted_code is not None:
+      file_resources.WriteReformattedCode(filename, reformatted_code, in_place,
+                                          encoding)
```
Comment (gwelymernans) [inline yapf/__init__.py]: Why did you move this down here? It no longer has the "`has_change`" check.

File: yapftests/main_test.py [hunk 8]
```diff
@@ -98,7 +98,6 @@ def testEchoBadInput(self):
     bad_syntax = "  a = 1\n"
     with patched_input(bad_syntax):
       with captured_output() as (out, err):
-        with self.assertRaisesRegexp(SyntaxError, "unexpected indent"):
           ret = yapf.main([])
```
Comment (gwelymernans) [inline yapftests/main_test.py]: Please dedent the `ret = yapf.main([])` line two spaces.

File: yapftests/yapf_test.py [hunk 9]
```diff
@@ -15,6 +15,7 @@
 """Tests for yapf.yapf."""
 
 import io
+from re import match
```
Comment (gwelymernans) [inline yapftests/yapf_test.py]: I prefer to use `import re` and use `re.match(...)` instead of importing `match`.

File: yapf/yapflib/yapf_api.py [hunk 10]
```diff
@@ -116,9 +121,21 @@ def FormatCode(unformatted_source,
...
+
+  fst_newline = unformatted_source.find('\n')
+
+  indent_match = re.match('^(\s+)', unformatted_source[:fst_newline])
+  st = style.CreateStyleFromConfig(style_config)
+  if indent_match:
+    unformatted_source = dedent(unformatted_source)
+    original_indent = indent_match.group(0)
+    original_len = len(original_indent)
+    offset = original_len - original_len % 4
```
Comment (gwelymernans) [inline yapf/yapflib/yapf_api.py]: What is the `4` in this line?

Comment (purpleP) [inline yapf/yapflib/yapf_api.py]: it's default number of spaces for one level indent in python.

File: yapftests/main_test.py (hunk 8 above)
Comment (purpleP) [inline yapftests/main_test.py]: Sure, I'll do that later today.

File: yapf/__init__.py (hunk 6 above)
Comment (purpleP) [inline yapf/__init__.py]: It should be, because if you add newline after joining you're creating a new string and when joining you don't

File: yapf/__init__.py (hunk 7 above)
Comment (purpleP) [inline yapf/__init__.py]: Probably I was just tired.

Comment (purpleP) [inline yapf/__init__.py]: What's more important is that it passes tests with that.

File: yapf/yapflib/yapf_api.py (hunk 10 above)
Comment (gwelymernans) [inline yapf/yapflib/yapf_api.py]: The number of spaces is adjustable. See the `INDENT_WIDTH` in `style.py`.

File: yapftests/yapf_test.py (hunk 9 above)
Comment (purpleP) [inline yapftests/yapf_test.py]: Unless you import a lot of things from module it's useless and just a waste of time and space in lines. So without a good reason I surely won't do that.

File: yapf/__init__.py (hunk 7 above)
Comment (gwelymernans) [inline yapf/__init__.py]: I believe the situation this was meant to solve was not writing to a file if there were no changes. Doing so updated the timestamp. It doesn't look like this is needed, though. So go ahead and move it back up please. :-)

File: yapftests/yapf_test.py (hunk 9 above)
Comment (gwelymernans) [inline yapftests/yapf_test.py]: How about "This is my project and I want it that way, and it keeps the namespace cleaner"?

File: yapf/yapflib/yapf_api.py (hunk 10 above)
Comment (purpleP) [inline yapf/yapflib/yapf_api.py]: so this should be style['INDENT_WITH'] right?

**Comment by purpleP** [comment]
@gwelymernans I've tried git rebase -i but I don't think it worked. Also the proposal about indenting that you've maid wasn't working.

File: yapftests/yapf_test.py (hunk 9 above)
Comment (purpleP) [inline yapftests/yapf_test.py]: How about you do that yourself then.

Comment (purpleP) [inline yapftests/yapf_test.py]: Like you have so polluted namespace anyway... It doesn't do any good, admit it =) I mean specifically in this case.

**Comment by gwelymernans** [comment]
@purpleP I'm not sure why there was a difficulty with rebasing, or even creating a new pull request. I'll close this out, because it's not really reviewable in its current form and also the CLA checks are now failing. :-(

**Comment by purpleP** [comment]
@gwelymernans I can create a new pull request if this would help. What is CLA?

**Comment by gwelymernans** [comment]
@purpleP Yeah, a new pull request would work best here. For the CLA, see the @googlebot comment above.
